pkg/util/slices.go

```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
 	}
 	return a
 }
+
+// Filter returns a new slice that only contains elements from collection that
+// satisfy predicate.
+//
+//	// Filter in place
+//	numbers = Filter(numbers, isEven)
+//	// Filter into a new slice
+//	odds := Filter(numbers, isEven)
+func Filter[T any](collection []T, predicate func(T) bool) []T {
+	i := 0
+	out := make([]T, len(collection))
+	for j := range collection {
+		if predicate(collection[j]) {
+			out[i] = collection[j]
+			i++
+		}
+	}
+	return slices.Clip(out[:i])
```

Comment:
The `Clip` seems unnecessary here, just return `out[:i]` would be sufficient.

The implementation of `Clip` [returns](https://cs.opensource.google/go/x/exp/+/92128663:slices/slices.go;l=403-405) the same slice using a full slice expression which forces any appends to reallocate it. Usually this is used to protect elements beyond the slice length that are "owned" by someone else. Since the slice is allocated within this function, and is passed entirely to the caller, there is nothing to protect. Moreover, the caller could save an allocation if they want to append something to this slice after the call; but `Clip` forces a reallocation in this case.